[READY] Bespoke Datum Mats

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

- Adds support for bespoke materials:
  - Reimplements [/datum/material/var/id] so VV can differentiate bespoke materials.
  - Ports GetIdFromArguments from SSdcs so that each bespoke material can have a unique id.
  - Adds a wrapper define for GetMaterialRef so people don't need to explicitly wrap the args in a list every time.
  - Adds [MATERIAL_INIT_BESPOKE] to differentiate bespoke materials and normal materials.
  - Adds [/datum/material/proc/Initialize] so bespoke materials can complain if they are passed invalid args
- Adds bespoke meats
  - Can be accessed by gibbing or butchering any mob that produces meat

- [x] TODO: Code docs
- [X] TODO: Actually adding bespoke materials
  - Added bespoke meats
- [x] TODO: Make butchering drop bespoke meats

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

- When I get around to it you will finally be able to tell who's bones/meat/skin that sofa is made from.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Bespoke materials exist
add: You can tell who's meat was used to make that chair
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros committed:
Bespoke Material Backend

- Adds support for bespoke materials:
  - Reimplements [/datum/material/var/id]
  - Ports GetIdFromArguments from SSdcs
  - Adds a wrapper define for GetMaterialRef
  - Adds [MATERIAL_INIT_BESPOKE]
  - Adds [/datum/material/proc/Initialize]
- Does not actually add any bespoke materials

- [ ] TODO: Code docs
- [ ] TODO: Actually adding bespoke materials

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

LemonInTheDark commented:
status?

TemporalOroboros commented:
I'm just about done with finals so I should be able to pick this back up in a day or two.

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

TemporalOroboros committed:
Some has_material procs and cleaning up some spaghetti

- Adds a pair of has_material procs for use in checking whether a given atom has a given material

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

TemporalOroboros committed:
Adds meat

- Adds bespoke meat variants
  - Does not make them accessible
- Shuts up the linter

TemporalOroboros committed:
Implements bespoke meat

- Makes the material container preserve bespoke materials
- Makes the sheetifier accept bespoke materials
- Makes the autolathe accept bespoke materials
- Makes the gibber produce bespoke meats

TemporalOroboros committed:
Makes butchering produce bespoke meats

This is jank and really needs to be folded into a unified butchering and gibbing system

Qustinnus commented on code/game/machinery/autolathe.dm:
```diff
@@ -123,18 +123,21 @@
 		use_power(min(1000, amount_inserted / 100))
 	updateUsrDialog()
 
+/// Checks whether a material is rigid
```
Why does this have to be handled through a callback? Seems like it should just be abled to handle that through the fact that we requested rigid mats to be accepted by it.

TemporalOroboros replied:
I'd love if it worked that way, but that's really not how material containers work. The material containers use the list of materials they contain to check whether a material fits into the container. If it isn't on the list, it doesn't fit. I could change that, but that would require more callbacks and more complicated code for just inserting objects.

If you want me to I'll do it.

TemporalOroboros replied:
On the other hand, reworking material containers like that would make instantiating new bespoke materials not O(n) in relation to the number of autolathes in existance.

TemporalOroboros replied:
Done

Qustinnus replied:
I think you're missing my point. We literaly tell it to use materials from SSmaterials.materialids_by_category[MAT_CATEGORY_RIGID]. This means we want it to accept all materials with MAT_CATEGORY_RIGID, but then we also have a callback which checks if the material is rigid. This means it's doing double work. Why not just make it so you pass MAT_CATEGORY_RIGID in the component creation and then the component checks if the material youre trying to put in has MAT_CATEGORY_RIGID. (Preferably, this would support lists so you could also say you only want materials that are both MAT_CATEGORY_RIGID and MAT_CATEGORY_ORE

TemporalOroboros replied:
I would like to be able to do this, the issue is that I have no idea what spaghetti expandable sets of materials future coders will want to have their machines to be able to contain. Ergo a callback for the cases where someone wants everything in three material categories that isn't in another two categories except actually contain this type but not these five for reasons.

Qustinnus replied:
that does not change that in most cases, you shouldnt have to make a callback. You can add a callback as an optional thing but it should not be required for every implementation that uses mat categories.

TemporalOroboros replied:
Done. The callback is now optional.

TemporalOroboros committed:
Material documentation

- Adds, fixes, and touches up some documentation

TemporalOroboros committed:
Material container insertion callback

- Changes the proc used to expand the material container's material list ot a proc used to check whether a material fits into a material container
- Instantiating new materials is no longer O(n) relative to the number of autolathes in existence.

TemporalOroboros committed:
Makes processing meat conserve materials

- Makes bespoke meat carry over into meatballs

TemporalOroboros committed:
Makes preserving custom materials an option

- Implements the ability to turn preserving custom materials _off_ for processor recipes

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into upstream/master

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

TemporalOroboros committed:
Fixes all bespoke materials of the same type using the same singleton

- We use ids now, not just types.

TemporalOroboros committed:
Makes the fat sucker produce bespoke meats

- Because consistency is good.

TemporalOroboros committed:
Fixes autolathes merging bespoke stacks into normal stacks.

TemporalOroboros committed:
Makes the callback to test materials for holdibility optional

- @Floyd

TemporalOroboros committed:
GetMaterialRef -> GET_MATERIAL_REF

- We capitalize macros.

Qustinnus commented on code/game/machinery/sheetifier.dm:
```diff
@@ -52,3 +53,7 @@
 	return ..()
+
+/// Checks whether a material is meat
+/obj/machinery/sheetifier/proc/CheckIsMeat(datum/material/meat/meat_ref)
```
This is not needed, is it?

TemporalOroboros replied:
No it is not, one sec.

TemporalOroboros replied:
Done!

TemporalOroboros committed:
Removes an extraneous callback

- Makes the sheetifier use functionality I didn't notice I implemented a few commits ago.

TemporalOroboros committed:
Makes mob and species meat null compatible

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

TemporalOroboros committed:
Fixes the ore silo

- The ore silo had really snowflake material handling that has been brought in line with the rest.
- The materials should show up in the correct order.

adamkad1 commented:
Ore silo havent got fixed tho

TemporalOroboros commented:
@adamkad1 I'm poking the people with the keys to get this updated.

TemporalOroboros committed:
Fixes minor lathe bugs

- Fixes stack_traces caused when lathes tried to fetch materials using reagent typepaths
- Fixed the selective reagent disposal topic. I have no idea how long this has been broken.

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

Jared-Fogle commented:
updated on terry

Qustinnus commented on code/controllers/subsystem/materials.dm:
```diff
@@ -31,40 +35,130 @@ SUBSYSTEM_DEF(materials)
+/// Returns a list to be used as an object's custom_materials. Lists will be cached and re-used based on the parameters.
 /datum/controller/subsystem/materials/proc/FindOrCreateMaterialCombo(list/materials_declaration, multiplier)
+	if(isnull(materials_declaration))
+		return // Shouldn't happen. Will happen.
```
throw an error then instead

TemporalOroboros replied:
Shouldn't this just pass back a null?

Qustinnus replied:
if this happens it means theres something going so doing CRASH() will make sure we can figure out why rather than having silent errors.

TemporalOroboros replied:
The problem is that there are legitimate reasons to pass this a null. Mostly clearing out somethings custom materials.

Qustinnus replied:
then the comment is wrong. it literaly says it shouldnt happen.

TemporalOroboros replied:
Fixed the comment

Qustinnus commented on code/__DEFINES/materials.dm:
```diff
@@ -9,6 +9,8 @@
 /// Used to make a material initialize at roundstart.
 #define MATERIAL_INIT_MAPLOAD	(1<<0)
+/// Used to make a material init bespoke.
+#define MATERIAL_INIT_BESPOKE	(1<<1)
```
This does not explain what being bespoke is to people who read the comment, maybe elaborate a bit more.

TemporalOroboros replied:
Done

Qustinnus reviewed (changes requested):
Fix these two things and it's good for merge. 

@LemonInTheDark I checked the code a few times and it's been TMd for quite a bit. Can you take a final look and / or merge it

LemonInTheDark commented on code/datums/components/material_container.dm:
```diff
@@ -10,42 +10,79 @@
 	precondition = _precondition
 	after_insert = _after_insert
 
+	for(var/mat in init_mats) //Make the assoc list ref | amount
```
what is the | trying to say here? Is it just like assoc list ref = value? You use || in the documentation earlier, is it meant to mean the same thing?

TemporalOroboros replied:
I'll clear it up.

TemporalOroboros replied:
Done

LemonInTheDark commented on code/datums/components/material_container.dm:
```diff
@@ -136,69 +173,106 @@
+	if(mat in allowed_materials)
+		return TRUE
+	if(istype(mat) && ((mat.id in allowed_materials) || (mat.type in allowed_materials)))
+		allowed_materials += mat	// This could get messy with passing lists by ref... but if you're doing that the list expansion is probably being taken care of elsewhere anyway...
```
You're making me nervous here man, how probable is probably?

LemonInTheDark replied:
Wait if the mat exists and it's id or it's type are already in the list, add it to the list?

TemporalOroboros replied:
I'm basically assuming that if someone is holding onto a reference of the allowed materials list they don't mind getting material datum instances in the list. The storing the datums is so I don't need to iterate over the list more than once if I get the same mat again.

LemonInTheDark commented on code/datums/components/material_container.dm:
```diff
@@ -136,69 +173,106 @@
+ * Arguments:
+ * - [mat][/atom/material]: The material we are checking for insertability.
+ */
+/datum/component/material_container/proc/can_hold_material(datum/material/mat)
+	if(mat in allowed_materials)
+		return TRUE
+	if(istype(mat) && ((mat.id in allowed_materials) || (mat.type in allowed_materials)))
+		allowed_materials += mat	// This could get messy with passing lists by ref... but if you're doing that the list expansion is probably being taken care of elsewhere anyway...
+		return TRUE
+	if(insertion_check?.Invoke(mat))
+		allowed_materials += mat
+		return TRUE
+	return FALSE
```
Early returns would be good here

TemporalOroboros replied:
I don't see what you mean, I'm already returning early.

LemonInTheDark replied:
brain empty no thoughts

LemonInTheDark commented on code/datums/components/material_container.dm:
```diff
@@ -10,42 +10,79 @@
-	if(allowed_types)
-		if(ispath(allowed_types) && allowed_types == /obj/item/stack)
-			allowed_typecache = GLOB.typecache_stack
+	allowed_materials = allowed_mats || list()
```
I've said before that I don't really like the overuse of this operator, but this isn't exactly easy to read in the first place, so why not

LemonInTheDark commented on code/datums/components/material_container.dm:
```diff
@@ -228,24 +302,25 @@
 
 	if(!target)
-		target = get_turf(parent)
+		var/atom/parent_atom = parent
+		target = parent_atom.drop_location()
```
can we be sure that parent will always be an atom?

TemporalOroboros replied:
Survey of the preexisting code says yes.

Qustinnus replied:
if you're only going to support atoms then make it check atom in Initialize.

TemporalOroboros replied:
Done

LemonInTheDark commented on code/datums/components/material_container.dm:
```diff
@@ -228,24 +302,25 @@
 	var/count = 0
 	while(sheet_amt > MAX_STACK_SIZE)
-		new M.sheet_type(target, MAX_STACK_SIZE)
+		new M.sheet_type(target, MAX_STACK_SIZE, null, list((M) = MINERAL_MATERIAL_AMOUNT))
```
excuse me?

TemporalOroboros replied:
This is so that the default material meat stacks don't assimilate the bespoke mat stacks when you break an autolathe. Stacks try to merge themselves into all other stacks of the same type and material in the loc they initialize in. Since meat stacks default to the normal meat material if there any partial stacks of the default meat material on the same turf (as will happen if an autolathe breaks) the bespoke mat stacks merge into the default mat stacks.

LemonInTheDark commented on code/datums/materials/meat.dm:
```diff
@@ -31,3 +31,23 @@
 	source.AddComponent(/datum/component/edible, list(/datum/reagent/consumable/nutriment = nutriment_count, /datum/reagent/consumable/cooking_oil = oil_count), null, RAW | MEAT | GROSS, null, 30, list("Fleshy"))
 
+
+/datum/material/meat/mob
+	init_flags = MATERIAL_INIT_BESPOKE
+
+/datum/material/meat/mob/Initialize(_id, mob/living/source)
+	if(!istype(source))
+		return FALSE
+
+	name = "[source?.name ? "[source.name]'s" : "mystery"] [initial(name)]"
+	return ..()
+
```
I don't like using keywords in typepaths like this, could you pick something else?

TemporalOroboros replied:
Sure, what would you recommend?

LemonInTheDark replied:
Maybe named? Honestly not too sure what would work well here

LemonInTheDark commented on code/modules/mob/living/simple_animal/hostile/goose.dm:
```diff
@@ -70,7 +70,7 @@
 	if(stat == DEAD || choking) // plapatin I swear to god
 		return FALSE
-	if(suffocator.custom_materials && suffocator.custom_materials[SSmaterials.GetMaterialRef(/datum/material/plastic)]) // dumb goose'll swallow food or drink with plastic in it
+	if(suffocator?.has_material_type(/datum/material/plastic)) // dumb goose'll swallow food or drink with plastic in it
```
why the ?. ? goose code ~~kill me~~ checks for the input atom quite often, did you see a runtime here before?

TemporalOroboros replied:
I suppose it isn't needed. I think I put that in on autopilot.

TemporalOroboros replied:
Done.

LemonInTheDark commented on code/modules/power/pipecleaners.dm:
```diff
@@ -234,10 +234,11 @@ By design, d1 is the smallest direction and d2 is the highest
 	. = ..()
 
-	color = param_color || color
+	if(param_color)
+		color = param_color
```
Thank you

LemonInTheDark commented on code/modules/research/designs.dm:
```diff
@@ -21,23 +21,41 @@ other types of metals and chemistry for reagents).
+	var/id = DESIGN_ID_IGNORE
+	/// Bitflags indicating what machines this design is compatable with. ([IMPRINTER]|[PROTOLATHE]|[AUTOLATHE]|[CRAFTLATHE]|[MECHFAB]|[BIOGENERATOR]|[LIMBGROWER]|[SMELTER]|[NANITE_COMPILER])
+	var/build_type = null
+	/// List of materials required to create one unit of the product. Format is typepath|category = amount
```
I assume the | here is meant to be or?

TemporalOroboros replied:
Yes. I'll clear it up.

LemonInTheDark commented on code/modules/research/machinery/_production.dm:
```diff
@@ -239,35 +265,47 @@
 /obj/machinery/rnd/production/proc/design_menu_entry(datum/design/D, coeff)
```
This procs output is cached, right?

TemporalOroboros replied:
It doesn't seem to be? It seems like it's just piped to the UI.

LemonInTheDark replied:
ah right, that's why the rnd machines cause so much overtime. I forgot

LemonInTheDark reviewed (changes requested):
Read it over, some questions about logic, mostly documentation stuff

TemporalOroboros commented on code/datums/components/material_container.dm:
```diff
@@ -10,42 +10,79 @@
 
 /// Sets up the proper signals and fills the list of materials with the appropriate references.
-/datum/component/material_container/Initialize(list/mat_list, max_amt = 0, _mat_container_flags=NONE, list/allowed_types, datum/callback/_precondition, datum/callback/_after_insert)
+/datum/component/material_container/Initialize(list/init_mats, max_amt = 0, _mat_container_flags=NONE, list/allowed_mats=init_mats, list/allowed_items, datum/callback/_insertion_check, datum/callback/_precondition, datum/callback/_after_insert)
```
Yep. You can define the default value of a proc argument to be pretty much anything, including vars from the same datum or other arguments. I'm not sure if the order matters. I've only ever defined later args off of earlier ones.

TemporalOroboros commented on code/datums/components/material_container.dm:
```diff
@@ -10,42 +10,79 @@
 	var/total_amount = 0
+	/// The maximum amount of materials this material container can contain
 	var/max_amount
+	/// Unused?
 	var/sheet_type
```
As far as I (read: VSCode reference finder) can tell they aren't used anywhere.

TemporalOroboros replied:
I'll remove them.

TemporalOroboros replied:
Removed

TemporalOroboros commented on code/datums/components/material_container.dm:
```diff
@@ -10,42 +10,79 @@
 	var/sheet_type
+	/// Map of material ref -> amount
 	var/list/materials //Map of key = material ref | Value = amount
+	/// Unused?
 	var/disable_attackby
```
Ditto

TemporalOroboros committed:
Various documentation fixes

- Clarified a couple comments
- Removes an extraneous ?. operator
- Fixed mat floor tiles having bugged reagent temperatures

TemporalOroboros committed:
More fixes

-/datum/material/meat/mob -> /datum/material/meat/mob_meat
- Adds atom typecheck to material containers.

Jared-Fogle commented:
CI is failing.

TemporalOroboros committed:
Merge branch 'master' into bespoke-mats

TemporalOroboros committed:
Fixes old typepaths
